test/unit/io/sequence_file/sequence_file_format_sam_test.cpp

```diff
@@ -0,0 +1,368 @@
+    expected_ids[2] = "ID3"; // "lala" is stripped
+    do_read_test(input);
+}
+
+TEST_F(read, no_qual)
+{
+  std::string input
+  {
+      "@ Comment\n"
+      "ID1\t0\t*\t0\t0\t*\t*\t0\t0\tACGTTTTTTTTTTTTTTT\t*\n"
+      "ID2\t0\t*\t0\t0\t*\t*\t0\t0\tACGTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTT"
+      "TTTTT\t!##$&'()*+,-./+)*+,-)*+,-)*+,-)*+,BDEBDEBDEBDEBDEBDEBDEBDEBDEBDEBDEBDEBDEBDEBDEBDE\n"
+      "ID3 lala\t0\t*\t0\t0\t*\t*\t0\t0\tACGTTTA\t!!!!!!!\n"
+    };
+
+    expected_quals[0] = { "*"_phred42 };
+    do_read_test(input);
+}
+
+TEST_F(read, wrong_qual)
+{
+    std::string input
+    {
+      "@ Comment\n"
+      "ID 1\t0\tBABABA\t200\t0\t*\tBABABA\t0\t0\tACGTTTTTTTTTTTTTTT\t!##$%&'()-./++-\tFI:i:1\n"
+    };
+
+    sequence_file_format_sam format;
+    sequence_file_input_options<dna5, false> options;
+    std::string id;
+    dna5_vector seq;
+    std::vector<phred42> qual;
+    std::stringstream istream{input};
+
+    EXPECT_THROW((format.read(istream, options, seq, id, qual)), parse_error );
+}
+
+TEST_F(read, wrong_qual2)
+{
+    std::string input
+    {
+      "@ Comment\n"
+      "ID 1\t0\tBABABA\t200\t0\t*\tBABABA\t0\t0\tACGTTTTTTTTTTTTTTT\t!##$%&'()*+,-./++-+\tFI:i:1\n"
+    };
+
+    sequence_file_format_sam format;
+    sequence_file_input_options<dna5, false> options;
+    std::string id;
+    dna5_vector seq;
+    std::vector<phred42> qual;
+    std::stringstream istream{input};
+
+    EXPECT_THROW((format.read(istream, options, seq, id, qual)), parse_error );
+}
+
+TEST_F(read, wrong_qual3)
+{
+    std::string input
+    {
+      "@ Comment\n"
+      "ID 1\t0\tBABABA\t200\t0\t*\tBABABA\t0\t0\tACGTTTTTTTTTTTTTTT\t!##$%&'()*\n,-./++-\tFI:i:1\n"
+    };
+
+    sequence_file_format_sam format;
+    sequence_file_input_options<dna5, false> options;
+    std::string id;
+    dna5_vector seq;
+    std::vector<phred42> qual;
+    std::stringstream istream{input};
+
+    EXPECT_THROW((format.read(istream, options, seq, id, qual)), parse_error );
+}
+
+TEST_F(read, no_id)
+{
+    std::string input
+    {
+      "@ Comment\n"
+      "*\t0\tBABABA\t200\t0\t*\tBABABA\t0\t0\tACGTTTTTTTTTTTTTTT\t!##$%&'()*+,-./++-\tFI:i:1\n"
+    };
+
+    sequence_file_format_sam format;
+    sequence_file_input_options<dna5, false> options;
+    std::string id;
+    dna5_vector seq;
+    std::vector<phred42> qual;
+    std::stringstream istream{input};
+
+    EXPECT_THROW((format.read(istream, options, seq, id, qual)), parse_error );
+}
+
+TEST_F(read, no_seq)
+{
+    std::string input
+    {
+      "@ Comment\n"
+      "ID 1\t0\tBABABA\t200\t0\t*\tBABABA\t0\t0\t*\t!##$%&'()*+,-./++-\tFI:i:1\n"
+    };
+
+    sequence_file_format_sam format;
+    sequence_file_input_options<dna5, false> options;
+    std::string id;
+    dna5_vector seq;
+    std::vector<phred42> qual;
+    std::stringstream istream{input};
+
+    EXPECT_THROW((format.read(istream, options, seq, id, qual)), parse_error );
```

Comment:
```suggestion
    EXPECT_THROW((format.read(istream, options, seq, std::ignore, std::ignore)), parse_error);
```

And then you can remove the variables. saves you some lines of code you do not need :) (check the no_id and no_wual as well)